File: dashboard/tests/simulation_engine.test.js
import test from 'node:test';
import assert from 'node:assert/strict';
import { SimulationMode, Presets, SimulationEngine, computeCrc16Ccitt } from '../src/simulation_engine.js';

test('SimulationEngine - generates signals for all attack modes', () => {
  const engine = new SimulationEngine();
  for (const mode of Object.values(SimulationMode)) {
    const res = engine.generateSignal(mode, null, 42, 256);
    assert.strictEqual(res.mode, mode);
    assert.strictEqual(res.expectedClass, mode);
    assert.strictEqual(res.frameLength, 29);
    assert.strictEqual(res.hexFrame.length, 58);
    assert.strictEqual(res.fftEnergyBins.length, 8);
    assert.strictEqual(res.voltageSamplesMv.length, 256);
    assert.strictEqual(res.crcValid, true);
  }
});

test('SimulationEngine - presets validation', () => {
  const engine = new SimulationEngine();
  for (const [name, preset] of Object.entries(Presets)) {
    const res = engine.generateSignal(preset.mode, preset.params, 42);
    assert.strictEqual(res.mode, preset.mode);
    assert.strictEqual(res.observedClass, preset.mode);
  }
});

test('SimulationEngine - deterministic seed produces identical signals', () => {
  const engine = new SimulationEngine();
  const res1 = engine.generateSignal(SimulationMode.EMP, Presets['Strong EMP'].params, 9999, 256, 1, 1000);
  const res2 = engine.generateSignal(SimulationMode.EMP, Presets['Strong EMP'].params, 9999, 256, 1, 1000);

  assert.strictEqual(res1.hexFrame, res2.hexFrame);
  assert.strictEqual(res1.crc16, res2.crc16);
  assert.deepStrictEqual(res1.fftEnergyBins, res2.fftEnergyBins);
  assert.deepStrictEqual(res1.voltageSamplesMv, res2.voltageSamplesMv);
});

test('SimulationEngine - different seed produces different noise', () => {
  const engine = new SimulationEngine();
  const res1 = engine.generateSignal(SimulationMode.NORMAL, Presets['Clean baseline'].params, 100);
  const res2 = engine.generateSignal(SimulationMode.NORMAL, Presets['Clean baseline'].params, 200);

  let diffFound = false;
  for (let i = 0; i < res1.voltageSamplesMv.length; i++) {
    if (Math.abs(res1.voltageSamplesMv[i] - res2.voltageSamplesMv[i]) > 0.001) {
      diffFound = true;
      break;
    }
  }
  assert.strictEqual(diffFound, true);
});

test('SimulationEngine - 29-byte frame encoding and CRC verification', () => {
  const engine = new SimulationEngine();
  const res = engine.generateSignal(SimulationMode.SURGE, Presets['Moderate grid surge'].params, 42);
  const bytes = [];
  for (let i = 0; i < res.hexFrame.length; i += 2) {
    bytes.push(parseInt(res.hexFrame.slice(i, i + 2), 16));
  }
  assert.strictEqual(bytes.length, 29);

  // Magic 0x5353
  assert.strictEqual(bytes[0], 0x53);
  assert.strictEqual(bytes[1], 0x53);

  // Recompute CRC over first 27 bytes
  const calculatedCrc = computeCrc16Ccitt(bytes, 27);
  const frameCrc = (bytes[27] << 8) | bytes[28];
  assert.strictEqual(calculatedCrc, frameCrc);
  assert.strictEqual(res.crc16, calculatedCrc);
});

test('SimulationEngine - alert gate behavior', () => {
  const engine = new SimulationEngine();
  const normalRes = engine.generateSignal(SimulationMode.NORMAL, null, 42);
  assert.strictEqual(normalRes.tamperDetected, false);

  const empRes = engine.generateSignal(SimulationMode.EMP, null, 42);
  assert.strictEqual(empRes.tamperDetected, true);
});

test('SimulationEngine - comparison and exports', () => {
  const engine = new SimulationEngine();
  const baseline = engine.generateSignal(SimulationMode.NORMAL, null, 42);
  const emp = engine.generateSignal(SimulationMode.EMP, null, 42);

  const comp = engine.compareSignals(baseline, emp);
  assert.strictEqual(comp.baselineClass, 'NORMAL');
  assert.strictEqual(comp.attackClass, 'EMP');
  assert.strictEqual(comp.spectralShift, 'High-Frequency RF');

  const json = engine.exportJson(emp);
  assert.ok(json.includes('"mode": "EMP"'));
  assert.ok(json.includes('"tamper_detected": true'));

  const csv = engine.exportCsv(emp);
  assert.ok(csv.includes('time_us,voltage_mv'));
});

File: android_app/app/src/test/java/com/sparkshield/android/simulation/SimulationEngineTest.kt
package com.sparkshield.android.simulation

import com.sparkshield.android.protocol.Crc16Ccitt
import com.sparkshield.android.protocol.TelemetryFrame
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test
import java.nio.ByteBuffer
import java.nio.ByteOrder

class SimulationEngineTest {

    private lateinit var engine: SimulationEngine

    @Before
    fun setUp() {
        engine = DefaultSimulationEngine()
    }

    @Test
    fun testAllModesGenerateSignals() {
        for (mode in SimulationMode.values()) {
            val params = when (mode) {
                SimulationMode.NORMAL -> NormalParameters()
                SimulationMode.EMP -> EmpParameters()
                SimulationMode.OPTICAL -> OpticalParameters()
                SimulationMode.SURGE -> SurgeParameters()
            }
            val result = engine.generateSignal(mode, params, seed = 100L)
            assertNotNull(result)
            assertEquals(mode, result.mode)
            assertEquals(mode.classLabel.name, result.expectedClass)
            assertEquals(29, result.frameLength)
            assertEquals(8, result.fftEnergyBins.size)
            assertTrue(result.confidence in 0.0f..1.0f)
            assertTrue(result.probabilities.size == 4)
        }
    }

    @Test
    fun testDefaultParameterPresets() {
        for (preset in SimulationPresets.allPresets) {
            val errors = preset.parameters.validate()
            assertTrue("Preset ${preset.name} had validation errors: $errors", errors.isEmpty())
            val res = engine.generateSignal(preset.mode, preset.parameters, seed = 42L)
            assertEquals(preset.mode.classLabel.name, res.expectedClass)
        }
    }

    @Test
    fun testParameterValidationRanges() {
        val invalidNormal = NormalParameters(lineFreqHz = 75.0f, baseAmplitudeMv = 1000.0f)
        val errors = invalidNormal.validate()
        assertTrue(errors.any { it.contains("50 or 60 Hz") })
        assertTrue(errors.any { it.contains("Base amplitude") })

        val invalidEmp = EmpParameters(peakVoltageMv = 5000.0f, riseTimeNs = 2.0f)
        val empErrors = invalidEmp.validate()
        assertTrue(empErrors.any { it.contains("peak voltage") })
    }

    @Test
    fun testDeterministicSeedProducesIdenticalSignals() {
        val params = EmpParameters(peakVoltageMv = 50000f, riseTimeNs = 25f)
        val res1 = engine.generateSignal(SimulationMode.EMP, params, seed = 12345L)
        val res2 = engine.generateSignal(SimulationMode.EMP, params, seed = 12345L)

        assertEquals(res1.hexFrame, res2.hexFrame)
        assertEquals(res1.crc16, res2.crc16)
        assertArrayEquals(res1.fftEnergyBins, res2.fftEnergyBins)
        assertEquals(res1.voltageSamplesMv.size, res2.voltageSamplesMv.size)
        for (i in res1.voltageSamplesMv.indices) {
            assertEquals(res1.voltageSamplesMv[i], res2.voltageSamplesMv[i], 0.0001f)
        }
    }

    @Test
    fun testDifferentSeedProducesDifferentNoise() {
        val params = NormalParameters()
        val res1 = engine.generateSignal(SimulationMode.NORMAL, params, seed = 111L)
        val res2 = engine.generateSignal(SimulationMode.NORMAL, params, seed = 999L)

        var hasDifference = false
        for (i in res1.voltageSamplesMv.indices) {
            if (kotlin.math.abs(res1.voltageSamplesMv[i] - res2.voltageSamplesMv[i]) > 0.001f) {
                hasDifference = true
                break
            }
        }
        assertTrue("Different seeds should produce different noise realizations", hasDifference)
    }

    @Test
    fun testFrameProtocolEncodingAndCrc() {
        val res = engine.generateSignal(SimulationMode.EMP, EmpParameters(), seed = 42L, sequenceId = 99L)
        assertEquals(29, res.frameLength)

        val bytes = res.hexFrame.chunked(2).map { it.toInt(16).toByte() }.toByteArray()
        assertEquals(29, bytes.size)

        val buffer = ByteBuffer.wrap(bytes).order(ByteOrder.BIG_ENDIAN)
        val magic = buffer.short.toInt() and 0xFFFF
        assertEquals(0x5353, magic)

        val seq = buffer.int
        assertEquals(99, seq)

        buffer.int // timestamp
        val flag = buffer.get().toInt() and 0xFF
        assertEquals(TelemetryFrame.FLAG_EMP, flag)

        val computedCrc = Crc16Ccitt.compute(bytes, 0, 27)
        buffer.position(27)
        val frameCrc = buffer.short.toInt() and 0xFFFF
        assertEquals(computedCrc, frameCrc)
        assertTrue(res.crcValid)
    }

    @Test
    fun testAlertGateBehavior() {
        // NORMAL mode: never triggers tamper alert
        val normalRes = engine.generateSignal(SimulationMode.NORMAL, NormalParameters(), seed = 42L)
        assertFalse("NORMAL mode must never trigger tamper alert", normalRes.tamperDetected)

        // EMP mode: triggers tamper alert when confidence >= 0.85
        val empRes = engine.generateSignal(SimulationMode.EMP, EmpParameters(), seed = 42L)
        if (empRes.confidence >= 0.85f && empRes.observedClass != "NORMAL") {
            assertTrue(empRes.tamperDetected)
        }
    }

    @Test
    fun testSignalComparisonAndShifts() {
        val baseline = engine.generateSignal(SimulationMode.NORMAL, NormalParameters(), seed = 42L)
        val emp = engine.generateSignal(SimulationMode.EMP, EmpParameters(), seed = 42L)

        val comp = engine.compareSignals(baseline, emp)
        assertEquals("NORMAL", comp.baselineClass)
        assertEquals("EMP", comp.attackClass)
        assertTrue(comp.deltaPeakMv > 0)
        assertEquals("High-Frequency RF", comp.spectralShift)
    }

    @Test
    fun testJsonAndCsvExport() {
        val res = engine.generateSignal(SimulationMode.OPTICAL, OpticalParameters(), seed = 42L)
        val json = engine.exportJson(res)
        assertTrue(json.contains("\"mode\": \"OPTICAL\""))
        assertTrue(json.contains("\"tamper_detected\":"))
        assertTrue(json.contains("\"safety_notice\":"))
        assertTrue(json.contains("\"crc16\":"))

        val csv = engine.exportCsv(res)
        assertTrue(csv.contains("time_us,voltage_mv"))
        assertTrue(csv.lines().size > 10)
    }
}
